nav2_collision_monitor/src/polygon.cpp

```diff
@@ -0,0 +1,376 @@
+}
+
+void Polygon::publish() const
+{
+  if (!visualize_) {
+    return;
+  }
+
+  auto node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  // Fill PolygonStamped struct
+  std::unique_ptr<geometry_msgs::msg::PolygonStamped> poly_s =
+    std::make_unique<geometry_msgs::msg::PolygonStamped>();
+  poly_s->header.stamp = node->now();
+  poly_s->header.frame_id = base_frame_id_;
+  poly_s->polygon = polygon_;
+
+  // Publish polygon
+  polygon_pub_->publish(std::move(poly_s));
+}
+
+bool Polygon::getCommonParameters(std::string & polygon_topic)
+{
+  auto node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  try {
+    // Get action type.
+    // Leave it not initialized: the will cause an error if it will not set.
+    nav2_util::declare_parameter_if_not_declared(
+      node, polygon_name_ + ".action_type", rclcpp::PARAMETER_STRING);
+    const std::string at_str =
+      node->get_parameter(polygon_name_ + ".action_type").as_string();
+    if (at_str == "stop") {
+      action_type_ = STOP;
+    } else if (at_str == "slowdown") {
+      action_type_ = SLOWDOWN;
+    } else if (at_str == "approach") {
+      action_type_ = APPROACH;
+    } else {  // Error if something else
+      RCLCPP_ERROR(logger_, "[%s]: Unknown action type: %s", polygon_name_.c_str(), at_str.c_str());
+      return false;
+    }
+
+    nav2_util::declare_parameter_if_not_declared(
+      node, polygon_name_ + ".max_points", rclcpp::ParameterValue(3));
+    max_points_ = node->get_parameter(polygon_name_ + ".max_points").as_int();
+
+    if (action_type_ == SLOWDOWN) {
+      nav2_util::declare_parameter_if_not_declared(
+        node, polygon_name_ + ".slowdown_ratio", rclcpp::ParameterValue(0.5));
+      slowdown_ratio_ = node->get_parameter(polygon_name_ + ".slowdown_ratio").as_double();
+    }
+
+    if (action_type_ == APPROACH) {
+      nav2_util::declare_parameter_if_not_declared(
+        node, polygon_name_ + ".time_before_collision", rclcpp::ParameterValue(2.0));
+      time_before_collision_ =
+        node->get_parameter(polygon_name_ + ".time_before_collision").as_double();
+      nav2_util::declare_parameter_if_not_declared(
+        node, polygon_name_ + ".simulation_time_step", rclcpp::ParameterValue(0.02));
+      simulation_time_step_ =
+        node->get_parameter(polygon_name_ + ".simulation_time_step").as_double();
+    }
+
+    nav2_util::declare_parameter_if_not_declared(
+      node, polygon_name_ + ".visualize", rclcpp::ParameterValue(false));
+    visualize_ = node->get_parameter(polygon_name_ + ".visualize").as_bool();
+    if (visualize_) {
+      // Get polygon topic parameter in case if it is going to be published
+      nav2_util::declare_parameter_if_not_declared(
+        node, polygon_name_ + ".polygon_topic", rclcpp::ParameterValue(polygon_name_));
+      polygon_topic = node->get_parameter(polygon_name_ + ".polygon_topic").as_string();
```

Comment:
Curious: does this also publish for the approach footprint?

Reply:
If set - yes. It might be useful for the `Circle` case, where we have no polygon footprint publisher. Since it is inside `getCommonParameters()` routine, this have no knowledge about what's base class calling it: polygon or circle. That is why there are no checks here. Hm... I am somewhere wrong: actually have, we could try to use `typeid(*this).name()`. But I do not want to have some hacking logic inside the code: this routine should not have a knowledge about any derived classes from it.

Reply:
Cool, that's a nice debugging element